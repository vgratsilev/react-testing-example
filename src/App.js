import { useEffect, useState } from 'react';
import Navbar from './components/Navbar/Navbar';
import AppRouter from './router/AppRouter';

function App() {
    const [data, setData] = useState(null);
    const [toggle, setToggle] = useState(false);
    const [value, setValue] = useState('');

    useEffect(() => {
        //const timeoutId =
        setTimeout(() => {
            setData({});
        }, 100);
        //return clearTimeout(timeoutId);
    }, []);

    const onClick = () => {
        setToggle((prev) => !prev);
    };

    return (
        <div>
            <div>
                <h1 data-testid={'value-elem'}>{value}</h1>
                {toggle && <div data-testid={'toggle-elem'}>Toggle</div>}
                {data && (
                    <div
                        id={'data-div'}
                        className={'mydiv'}
                    >
                        hided element
                    </div>
                )}
                <h1>Test 1</h1>
                <button
                    type={'button'}
                    data-testid={'toggle-btn'}
                    onClick={onClick}
                >
                    Click
                </button>
                <input
                    type={'text'}
                    placeholder={'input value...'}
                    value={value}
                    onChange={(e) => setValue(e.target.value)}
                />
            </div>
            <div>
                <Navbar />
                <AppRouter />
            </div>
        </div>
    );
}

export default App;
