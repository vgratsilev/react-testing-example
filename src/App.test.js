/* eslint-disable testing-library/no-debugging-utils */
import { fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';
import { renderWithRouter } from './tests/helpers/renderWithRouter';

describe('TEST APP', () => {
    test('renders', () => {
        renderWithRouter(<App />);
        const h1 = screen.getByText(/Test 1/i);
        const button = screen.getByRole('button');
        const input = screen.getByPlaceholderText(/input value/i);
        expect(h1).toBeInTheDocument();
        expect(button).toBeInTheDocument();
        expect(input).toMatchSnapshot();
        //screen.debug();
    });

    test('Async showed element', async () => {
        renderWithRouter(<App />);
        // const h1 = screen.queryByText(/hello2/i);
        // expect(h1).toBeNull();

        const elem = await screen.findByText('hided element');
        // expect(elem).toBeNull();
        expect(elem).toBeInTheDocument();
        expect(elem).toHaveClass('mydiv');
        screen.debug();
    });

    test('Btn Click event', () => {
        renderWithRouter(<App />);
        const btn = screen.getByTestId('toggle-btn');
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
        fireEvent.click(btn);
        // eslint-disable-next-line testing-library/prefer-presence-queries
        expect(screen.queryByTestId('toggle-elem')).toBeInTheDocument();
        fireEvent.click(btn);
        expect(screen.queryByTestId('toggle-elem')).toBeNull();
    });

    test('Input synthetic event', () => {
        renderWithRouter(<App />);
        const input = screen.getByPlaceholderText(/input value/i);
        expect(screen.queryByTestId('value-elem')).toContainHTML('');

        // synthetic event
        fireEvent.input(input, {
            target: { value: '123456' },
        });
        expect(screen.queryByTestId('value-elem')).toContainHTML('123456');
    });

    test('Input user event', () => {
        renderWithRouter(<App />);
        const input = screen.getByPlaceholderText(/input value/i);
        expect(screen.queryByTestId('value-elem')).toContainHTML('');

        // user event
        userEvent.type(input, '123456');
        expect(screen.queryByTestId('value-elem')).toContainHTML('123456');
    });
});
