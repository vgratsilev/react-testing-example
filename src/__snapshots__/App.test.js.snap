// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST APP renders 1`] = `
<input
  placeholder="input value..."
  type="text"
  value=""
/>
`;
